src/Orleans/ApplicationParts/AssemblyAttributeFeatureProvider.cs

```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Orleans.Metadata;
+
+namespace Orleans.ApplicationParts
+{
+    /// <summary>
+    /// Populates a feature using assembly-level attributes which implement <see cref="IFeaturePopulator{TFeature}"/>.
+    /// </summary>
+    /// <typeparam name="TFeature">The feature type.</typeparam>
+    public sealed class AssemblyAttributeFeatureProvider<TFeature> : IApplicationFeatureProvider<TFeature>
+    {
+        /// <summary>
+        /// Returns <see langword="true"/> if this instance is equivalent to the provided instance, <see langword="false"/> otherwise.
+        /// </summary>
+        /// <param name="other">The other instance.</param>
+        /// <returns>
+        /// <see langword="true"/> if this instance is equivalent to the provided instance, <see langword="false"/> otherwise.
+        /// </returns>
+        public bool Equals(AssemblyAttributeFeatureProvider<TFeature> other) => true;
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            return obj.GetType() == this.GetType();
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            return typeof(AssemblyAttributeFeatureProvider<TFeature>).GetHashCode();
+        }
+
+        /// <inheritdoc />
+        public void PopulateFeature(IEnumerable<IApplicationPart> parts, TFeature feature)
+        {
+            foreach (var part in parts.OfType<AssemblyPart>())
+            {
+                var attributes = part.Assembly.GetCustomAttributes(true);
+                foreach (var attribute in attributes.OfType<FeaturePopulatorAttribute>())
+                {
+                    if (!typeof(IFeaturePopulator<TFeature>).IsAssignableFrom(attribute.PopulatorType)) continue;
```

Comment:
Why not fail? Silent skipping can be misleading for developers who adds their own features, not?

Reply:
Because not all attributes in an assembly are feature populators and the ones that are might not be feature populators for this particular type.

Reply:
We are only iterating on ```FeaturePopulatorAttribute```s, so first part of your reply is is covered,  what other types are we supporting on the ```PopulatorType``` property which ones are not ```IFeaturePopulator<TFeature>```?